src/transaction/transaction_manager.cpp

```diff
@@ -5,21 +5,40 @@
 
 namespace terrier::transaction {
 TransactionContext *TransactionManager::BeginTransaction() {
-  // This latch has to also protect addition of this transaction to the running transaction table. Otherwise,
-  // the thread might get scheduled out while other transactions commit, and the GC will deallocate their version
-  // chain which may be needed for this transaction, assuming that this transaction does not exist.
-  common::SharedLatch::ScopedSharedLatch guard(&commit_latch_);
-  timestamp_t start_time = time_++;
+  timestamp_t start_time;
+  {
+    // There is a three-way race that needs to be prevented.  Specifically, we
+    // cannot allow both a transaction to commit and the GC to poll for the
+    // oldest running transaction in between this transaction acquiring its
+    // begin timestamp and getting inserted into the current running
+    // transactions list.  Using the current running transactions latch
+    // prevents the GC from polling and stops the race.  This allows us to
+    // replace acquiring a shared instance of the commit latch with a
+    // read-only spin-latch and move the allocation out of a critical section.
+    common::SpinLatch::ScopedSpinLatch running_guard(&curr_running_txns_latch_);
+    start_time = time_++;
 
-  // TODO(Tianyu):
-  // Maybe embed this into the data structure, or use an object pool?
-  // Doing this with std::map or other data structure is risky though, as they may not
-  // guarantee that the iterator or underlying pointer is stable across operations.
-  // (That is, they may change as concurrent inserts and deletes happen)
+    // TODO(Tianyu):
+    // Maybe embed this into the data structure, or use an object pool?
+    // Doing this with std::map or other data structure is risky though, as they may not
+    // guarantee that the iterator or underlying pointer is stable across operations.
+    // (That is, they may change as concurrent inserts and deletes happen)
+    const auto ret UNUSED_ATTRIBUTE = curr_running_txns_.emplace(start_time);
+    TERRIER_ASSERT(ret.second, "commit start time should be globally unique");
+  }  // Release latch on current running transactions
+
+  // Do the allocation outside of any critical section
   auto *const result = new TransactionContext(start_time, start_time + INT64_MIN, buffer_pool_, log_manager_, this);
-  common::SpinLatch::ScopedSpinLatch running_guard(&curr_running_txns_latch_);
-  const auto ret UNUSED_ATTRIBUTE = curr_running_txns_.emplace(result->StartTime());
-  TERRIER_ASSERT(ret.second, "commit start time should be globally unique");
+
+  // This transaction cannot proceed if there is an ongoing write-commit by a
+  // transaction whose commit time is less than this transaction's begin time
+  //
+  // TODO(John):  This check could be replaced with a timestamp comparison to
+  // potentially allow some transactions who may be blocked unnecessarily to
+  // proceed.
+  while (blocking_commit_.load()) {
+  }
```

Comment:
Should you use `_mm_pause()` or whatever the fancy new Intel thing is (`__asm__ __volatile__("pause;")` in tbb? check their source code)?

Reply:
A quick test of this slowed down the benchmarks by 2-3%.